src/backend/libpq/rangerrest.c

```diff
@@ -281,29 +283,27 @@ static size_t write_callback(char *contents, size_t size, size_t nitems,
 	CURL_HANDLE curl = (CURL_HANDLE) userp;
 	Assert(curl != NULL);
 
-	if (curl->response.buffer == NULL) 
-	{
-		curl->response.buffer = palloc0(realsize + 1);
-	}
-	else 
+	elog(DEBUG3, "response size is %d. response buffer size is %d.", curl->response.size, curl->response.buffer_size);
+	if(curl->response.size + realsize >= curl->response.buffer_size)
 	{
-		/*Note:*/
-		/*our repalloc is not same as realloc, repalloc's first param(buffer) can not be NULL*/
-		curl->response.buffer = repalloc(curl->response.buffer, curl->response.size + realsize + 1);
+		/*
+		 * our repalloc is not same as realloc, repalloc's first param(buffer) can not be NULL
+		 * double the buffer size if the buffer is not enough.
+		 */
+		curl->response.buffer = repalloc(curl->response.buffer, curl->response.buffer_size * 2);
```

Comment:
Oh, buffer*2 maybe not enough for current content.
Should use a loop to determine the next buffer_size:
```
while (curl->response.size + realsize >= curl->response.buffer_size)
{ curl->response.buffer_size *= 2; }
//do realloc
//...
```